fetch_index.py

```diff
@@ -130,237 +218,320 @@ def index_lists(self, sites, collection, index):
+        """This method when invoked would index the permission provided in the paramater
+            for the user in paramter user_name
+            :param user_name: a string value denoting the username of the user
+            :param permission: permission that needs to be provided to the user
+        """
+        try:
+            self.ws_client.add_user_permissions(
+                content_source_id=self.ws_source,
+                http_auth=self.ws_token,
+                user=user_name,
+                body={
+                    "permissions": [permission]
+                },
             )
-            roles = self.sharepoint_client.get(
-                host_url,
-                "web/roleassignments?$expand=Member/users,RoleDefinitionBindings",
+            logger.info(
+                "Successfully indexed the permissions for user %s to the workplace" % (
+                    user_name
+                )
+            )
+        except Exception as exception:
+            logger.exception(
+                "Error while indexing the permissions for user: %s to the workplace. Error: %s" % (
+                    user_name, exception
+                )
             )
-        roles = roles.json()
+            self.is_error = True
+            return []
+
+    def index_permissions(
+        self,
+        key,
+        collection,
+        site=None,
+        list_name=None,
+        list_url=None,
+        itemid=None,
+    ):
+        """This method when invoked, checks the permission inheritance of each object.
+            If the object has unique permissions, the list of users having access to it
+            is fetched using sharepoint api else the permission levels of the that object
+            is taken same as the permission level of the site collection.
+            :param key: key, a string value
+            :param collection: collection name
+            :param site: site name to index the permission for the site
+            :param list_name: list name to index the permission for the list
+            :param list_url: url of the list
+            :param itemid: item id to index the permission for the item
+            Returns:
+                groups: list of users having access to the given object
+        """
+        roles, rel_url = self.get_roles(key, site, list_url, list_name, itemid)
+
         groups = []
 
-        roles = roles.get("d", {}).get("results")
+        roles = check_response(roles, "Cannot fetch the roles for the given object %s at url %s" % (
+            key, rel_url), "Error while parsing response for fetch_users for %s at url %s." % (key, rel_url), "roles")
 
+        rows = {}
+        if (os.path.exists(self.mapping_sheet_path) and os.path.getsize(self.mapping_sheet_path) > 0):
+            with open(self.mapping_sheet_path) as file:
+                csvreader = csv.reader(file)
+                for row in csvreader:
+                    rows[row[0]] = row[1]
         for role in roles:
```

Comment:
No, only the groups will be indexed as permissions for the users.
However, you're right, permission indexing has to follow indexing schedule. The official docs states that permission indexing can only occur at the time of indexing. We are still looking for a better approach. Do let me know if you have any other approach.

Reply:
We would have to explore why can't permissions run independently because the permissions does not have an update method.
Hence we will have to
1. Fetch in latest permissions for all the IDs indexed in ws
2. Delete all the permissions
3. Index new permissions
Which is currently being taken care by fullsync

Reply:
Hi @ioanatia , seems for custom source, there's no external entities mapping. So how should a connector initialize a `permissions` job?.